### PR TITLE
Split items in `misc` directory by topic

### PR DESCRIPTION
**Pull request by aidecoe**

### CODE REVIEW

**Comment by marmarek** [comment]
Summary of IRC discussion:
* Don't segregate into directories based on file type, but rather on their function.
* Creating a directory for 1-2 files makes navigation harder, not easier.

Based on the above, some ideas for organizing `misc` into more meaningful directories:
 - `passwordless-root` - sudo, polkit and pam configuration for this (note `sudoers.d_umask` and `sudoers.d_qt_x11_no_mitshm` are _not_ related to passwordless root)
 - `updates` or `package-managers` or something like this - files related to accessing package repositories: repository definitions, extra configuration for package managers, making them use updates proxy, checking updates availability

**Comment by aidecoe** [comment]
### passwordless-root

That seems most straightforward:

```
$ ls -1 passwordless-root/
Makefile
pam.d_su.qubes
polkit-1-qubes-allow-all.pkla
polkit-1-qubes-allow-all.rules
qubes.sudoers
```

### package manager related

I think it would still be worth to split the below into per distro directory, at least to spare reader thinking which of the below files are relevant.
```
$ ls -1 package-managers/
apt-conf-70no-unattended
dnf-qubes-hooks.conf
dnf-qubes-hooks.py
Makefile
qubes-download-dom0-updates.sh
qubes-r4.list.in
qubes-r4.repo.in
RPM-GPG-KEY-qubes-4-centos
RPM-GPG-KEY-qubes-4-primary
RPM-GPG-KEY-qubes-4-unstable
upgrades-installed-check
upgrades-status-notify
yum-qubes-hooks.conf
yum-qubes-hooks.py
```

**Comment by aidecoe** [comment]
Before I'd get rid of `helpers`, how would we partition the scripts below?

```
$ ls -1 helpers/
close-window.c
Makefile
qubes-desktop-run
qubes-run-terminal
qubes-run-terminal.desktop
qubes-session-autostart
qubes-trigger-sync-appmenus.sh
qvm-connect-tcp
qvm-console
qvm-features-request
resize-rootfs
tinyproxy-wrapper
xenstore-watch.c
```

**Comment by marmarek** [comment]
> $ ls -1 helpers/
> close-window.c

I'm pretty sure it isn't used anymore.

> Makefile

> qubes-desktop-run
> qubes-run-terminal
> qubes-run-terminal.desktop
> qubes-session-autostart
> qubes-trigger-sync-appmenus.sh

GUI related? application-menu-scripts?

> qvm-connect-tcp

network?

> qvm-console

TBH, this would better fit core-admin-client repository.

> qvm-features-request
> resize-rootfs

> tinyproxy-wrapper

init? or verify if still needed and perhaps drop?

> xenstore-watch.c

not needed anymore

**Comment by marmarek** [comment]
Both Debian and Fedora build fails (see travis output). But besides those technical issues, generally looks good now. Let me know when it will be ready for the final review, or when you need some input from me.

**Comment by aidecoe** [comment]
Thanks, Marek. I'll let you know.

**Comment by aidecoe** [comment]
@marmarek is `misc/50_qubes.conf` config for needrestart in Debian still relevant? isn't it outdated by any chance?

**Comment by marmarek** [comment]
> @marmarek is `misc/50_qubes.conf` config for needrestart in Debian still relevant? isn't it outdated by any chance?

TBH not sure about it @adrelanos ?
I mean, those services shouldn't be restarted on package update automatically, but I don't know what is the state of needrestart in Debian.

**Comment by adrelanos** [comment]
https://github.com/QubesOS/qubes-core-agent-linux/commit/7dc99ee66250834df4ead3709ac4d2371a7ce5ef seems still relevant. Nothing related changed in needrestart.

**Comment by aidecoe** [comment]
@adrelanos in the line

    nrconf{override_rc}->{q(^qubes-core-agent-linux)} = 0;

is `qubes-core-agent-linux` a package name? There's no `qubes-core-agent-linux` for Debian, only `qubes-core-agent` (without `-linux` suffix).

**Comment by adrelanos** [comment]
It's a systemd unit file (service) name.

https://github.com/liske/needrestart/blob/10bd2db5e2d88370882cb74d729392f5a1389098/ex/needrestart.conf#L65

Should probably be `qubes-core-agent` indeed.

**Comment by aidecoe** [comment]
@marmarek what is `misc/udev-qubes-misc.rules` for?

**Comment by marmarek** [comment]
> @marmarek what is `misc/udev-qubes-misc.rules` for?

Makes memory hotplug working (for dynamic memory management)

**Comment by aidecoe** [comment]
In rpm spec:

```
mkdir -p /var/lib/qubes
if [ -e /etc/fstab ] ; then
mv /etc/fstab /var/lib/qubes/fstab.orig
fi
```

- Doesn't this keep overriding `fstab.orig` on every install? If so, what's the point?

**Comment by marmarek** [comment]
No, this part (see part just above it) is done only on the first install.

**Comment by aidecoe** [comment]
Most of the stuff is moved. I'm not sure about app-defaults and sys-defaults - I think I need to find a better name or place for those.

init.d and systemd - those I'll address in later Pull Request.

File: Makefile [hunk 1]
```diff
@@ -52,8 +47,8 @@ clean:
 	rm -f .coverage
 
 all:
-	make -C misc
-	make -C qubes-rpc
+	$(MAKE) -C misc VERSION=$$(< version)
```
Comment (marmarek) [inline Makefile]: You already have `$(VERSION)`, you can use that here, instead of re-reading the file.

File: debian/qubes-core-agent.install [hunk 2]
```diff
@@ -51,7 +51,7 @@ etc/sudoers.d/umask
 etc/sysctl.d/20_tcp_timestamps.conf
 etc/sysctl.d/80-qubes.conf
 etc/systemd/system/haveged.service
-etc/udev/rules.d/50-qubes-misc.rules
+etc/udev/rules.d/50-qubes-mem-hotplug.rules
```
Comment (marmarek) [inline debian/qubes-core-agent.install]: While at it, this should really be in `/lib/udev`, `/etc` is for user configs.

File: rpm_spec/core-agent.spec.in [hunk 3]
```diff
@@ -310,12 +324,6 @@ fi
 
 %triggerin -- grub2-tools
 
-if ! grep -q /etc/default/grub.qubes /etc/default/grub 2>/dev/null; then
-    # do not keep Qubes-related settings directly in user-controlled config,
-    # include another file
-    echo '. /etc/default/grub.qubes' >> /etc/default/grub
-fi
```
Comment (marmarek) [inline rpm_spec/core-agent.spec.in]: Where did this go? From what I know, Fedora don't support `/etc/default/grub.d` yet, so sadly we still need this hack.

File: rpm_spec/core-agent.spec.in [hunk 4]
```diff
@@ -575,7 +583,7 @@ rm -f %{name}-%{version}
 %config(noreplace) /etc/qubes/rpc-config/qubes.InstallUpdatesGUI
 %config(noreplace) /etc/qubes/rpc-config/qubes.VMShell+WaitForSession
 %dir /etc/qubes/autostart
-%config(noreplace) /etc/default/grub.qubes
+%config(noreplace) /etc/grub.d/30_qubes
```
Comment (marmarek) [inline rpm_spec/core-agent.spec.in]: `/etc/grub.d` is not the same as `/etc/default/grub.d`. And the latter is AFAIR Debian-only.

File: boot/Makefile [hunk 5]
```diff
@@ -0,0 +1,10 @@
+LIBDIR ?= /usr/lib
+SYSCONFDIR ?= /etc
+GRUBCONFDIR ?= $(SYSCONFDIR)/grub.d
+DRACUTCONFDIR ?= $(LIBDIR)/dracut/dracut.conf.d
+
+.PHONY: install
+
+install:
+	install -D -m 0644 grub.qubes $(DESTDIR)$(GRUBCONFDIR)/30_qubes
```
Comment (marmarek) [inline boot/Makefile]: This needs to go into `/etc/default/grub.d` on Debian. And somewhere similar (but to be included manually) on Fedora.

**Comment by aidecoe** [comment]
> The `qvm-console` move needs a PR actually adding it into the other repo.

@marmarek https://github.com/QubesOS/qubes-core-admin-client looks like a Python package. Are you sure it's a good place to move `qvm-console` to? I think it doesn't belong there either...

File: Makefile (hunk 1 above)
Comment (aidecoe) [reply Makefile]: fixed

File: debian/qubes-core-agent.install (hunk 2 above)
Comment (aidecoe) [reply debian/qubes-core-agent.install]: done

**Comment by marmarek** [comment]
> > The `qvm-console` move needs a PR actually adding it into the other repo.
> 
> @marmarek https://github.com/QubesOS/qubes-core-admin-client looks like a Python package. Are you sure it's a good place to move `qvm-console` to? I think it doesn't belong there either...

It is a mostly python package, yes. But functionally it is also a logical place for such tool.

File: rpm_spec/core-agent.spec.in (hunk 3 above)
Comment (aidecoe) [reply rpm_spec/core-agent.spec.in]: I have restored it.

File: boot/Makefile (hunk 5 above)
Comment (aidecoe) [reply boot/Makefile]: Please see `debian/rules`.

File: rpm_spec/core-agent.spec.in (hunk 4 above)
Comment (aidecoe) [reply rpm_spec/core-agent.spec.in]: I have restored it.

**Comment by adrelanos** [comment]
> +	install -D -m 0644 grub.qubes $(DESTDIR)$(GRUBCONFDIR)/30_qubes


Doesn't this need to end with file extension '.cfg'? All files in
/etc/default/grub.d end with '.cfg'.

**Comment by adrelanos** [comment]
Amadeusz Żołnowski:
> -    echo '. /etc/default/grub.qubes' >> /etc/default/grub


Ideally please don't write to /etc/default/grub file and exclusively use
/etc/default/grub.d folder.

Related:

Deduplicate /boot/grub/grub.cfg kernel boot parameters /etc/default/grub

https://github.com/QubesOS/qubes-issues/issues/5340

**Comment by marmarek** [comment]
> Ideally please don't write to /etc/default/grub file and exclusively use /etc/default/grub.d folder.

In Debian - sure, this is what we do. But the package in Fedora doesn't support `/etc/default/grub.d` directory.

**Comment by aidecoe** [comment]
> + install -D -m 0644 grub.qubes $(DESTDIR)$(GRUBCONFDIR)/30_qubes
> Doesn't this need to end with file extension '.cfg'? All files in /etc/default/grub.d end with '.cfg'.

Generally, not necessarily:
https://www.gnu.org/software/grub/manual/grub/grub.html#Simple-configuration

But distro specific, e.g. if /etc/default/grub.d is used instead, perhaps. At the end I have realized that it's impossible to make this part generic, so Makefile follows defaults per tool and rpm / deb make adjustments.

**Comment by aidecoe** [comment]
> The `qvm-console` move needs a PR actually adding it into the other repo.

See See QubesOS/qubes-core-admin-client#131 where qvm-console is removed from qubes-core-agent-linux.

File: rpm_spec/core-agent.spec.in [hunk 6]
```diff
@@ -586,9 +597,11 @@ rm -f %{name}-%{version}
 %dir /etc/qubes/post-install.d
 /etc/qubes/post-install.d/README
 /etc/qubes/post-install.d/*.sh
+%config(noreplace) /etc/profile.d/qt_x11_no_mitshm.sh
 %config(noreplace) /etc/sudoers.d/qt_x11_no_mitshm
+%config(noreplace) /etc/sudoers.d/umask
 %config(noreplace) /etc/sysctl.d/20_tcp_timestamps.conf
-%config(noreplace) /etc/udev/rules.d/50-qubes-misc.rules
+%config(noreplace) /lib/udev/rules.d/50-qubes-mem-hotplug.rules
```
Comment (marmarek) [inline rpm_spec/core-agent.spec.in]: This should no longer be marked as (used-editable) config.

File: misc/close-window.c [hunk 7]
```diff
@@ -1,42 +0,0 @@
-#include <X11/Xutil.h>
```
Comment (marmarek) [inline misc/close-window.c]: Removing files should be in separate commit (with reasoning), not with one about moving files without changes.

File: misc/dconf-profile-user [hunk 8]
```diff
@@ -1,2 +0,0 @@
-user-db:user
```
Comment (marmarek) [inline misc/dconf-profile-user]: And here too (move to separate commit with explanation).

File: misc/xenstore-watch.c [hunk 9]
```diff
@@ -1,29 +0,0 @@
-#include <sys/types.h>
```
Comment (marmarek) [inline misc/xenstore-watch.c]: And here too.

**Comment by marmarek** [comment]
There are some side effects - `qubes-core-agent` rpm got extra:
 - `/etc/profile.d/qt_x11_no_mitshm.sh`
 - `/etc/sudoers.d/umask`

Previously they were included in deb only. But it rather looks like the previous situation was wrong.

**Comment by aidecoe** [comment]
> There are some side effects - `qubes-core-agent` rpm got extra:
> 
>     * `/etc/profile.d/qt_x11_no_mitshm.sh`
> 
>     * `/etc/sudoers.d/umask`
> 
> 
> Previously they were included in deb only. But it rather looks like the previous situation was wrong.

Yes, that was intentional. Does it need separate commit? Or perhaps mention in commit message would be enough (preferred (-:)?

**Comment by marmarek** [comment]
From IRC discussion: I'm not fully convinced for moving `qubesagent/` to `qubesagent/qubesagent/`. The only thing it does is reduce number of top level files (`setup.py`, `run-tests`, `test-packages`), without clear improvement in semantics of the directory. At the same time, introduce non-standard location for setup.py and python module subdir - not a big issue, but changes development workflow a little (for example you can't run just `python3 -m unittest` from the top directory anymore).

Reducing top-level files can be achieved differently: integrate `run-tests` into Makefile, move `test-packages` into `ci` subdir. Then at the top-level, python stuff will be just `setup.py`, and `qubesagent` (and `ci`, but it can be considered generic, even though only python-tests files are there right now).

Regardless of the above, it would really help if you add README.md with explanation what goes into what directory - to avoid the current mess in the future.

**Comment by aidecoe** [comment]
> Regardless of the above, it would really help if you add README.md with explanation what goes into what directory - to avoid the current mess in the future.

OK, will add.

**Comment by aidecoe** [comment]
> [...] moving `qubesagent/` to `qubesagent/qubesagent/`. The only thing it does is reduce number of top level files (`setup.py`, `run-tests`, `test-packages`), without clear improvement in semantics of the directory.

It's not about number of top level files at all. It is about separating different topics. I see Python qubesagent package as a separate thing. Unless... this is actually the agent (as in name of the repo: qubes-core-agent-linux) and most of the rest accompanying directories belong to a separate project/repo (like certainly qubes-rpc does). In such case we certainly shall keep setup.py in top level and move the stuff that is not strictly required for qubesagent out of this repo.

> At the same time, introduce non-standard location for setup.py and python module subdir - not a big issue, but changes development workflow a little (for example you can't run just `python3 -m unittest` from the top directory anymore).

But if you're working on qubesagent Python stuff, you cd to that dir and carry on. If you work on e.g. package managers integration, you cd to package-managers.

**Comment by aidecoe** [comment]
> > Regardless of the above, it would really help if you add README.md with explanation what goes into what directory - to avoid the current mess in the future.
> 
> OK, will add.

Please have a look at `package-managers/README.md` whether it's going in right direction.

**Comment by marmarek** [comment]
This readme looks good, but honestly I had in mind a README file at top level with 1-2 sentences about each directory. In addition to describing directory split, it will also serve as description what is in the repository.

**Comment by aidecoe** [comment]
@marmarek what about qubesagent subdir? Could you refer to my reply above, please?

Can README.md files wait and we focus on merging this? Recently I keep rebasing and resolving conflicts again and again.

**Comment by marmarek** [comment]
qubesagent dir is really a convenient packaging of python scripts that serve different purposes, so they can be imported from different places. It's cumbersome (at least) to import python files that aren't installed as modules. Otherwise, different files there would live together with relevant stuff (`*firewall.py` inside `network/`, `xdg.py` inside `app-menu/`). Theoretically we could have separate python packages for each (and then inside relevant category dir), that could be a cleaner solution.
Would you be willing to split it, create separate setup.py, adjust imports and packaging?

**Comment by marmarek** [comment]
> Can README.md files wait and we focus on merging this? Recently I keep rebasing and resolving conflicts again and again.

ok

**Comment by aidecoe** [comment]
> qubesagent dir is really a convenient packaging of python scripts that serve different purposes, so they can be imported from different places. It's cumbersome (at least) to import python files that aren't installed as modules. Otherwise, different files there would live together with relevant stuff (`*firewall.py` inside `network/`, `xdg.py` inside `app-menu/`). Theoretically we could have separate python packages for each (and then inside relevant category dir), that could be a cleaner solution.
> Would you be willing to split it, create separate setup.py, adjust imports and packaging?

That makes sense. I'll try that in a follow-up PR. What do we in the meantime with this PR? Revert back to top level (some effort me) or keep my change?

**Comment by marmarek** [comment]
> That makes sense. I'll try that in a follow-up PR. What do we in the meantime with this PR? Revert back to top level (some effort me) or keep my change?

Please revert. I think it should be enough to rebase -i and skip this one commit.

**Comment by marmarek** [comment]
I'm holding off merging other PRs until this one is one is merged, to prevent further conflicts.

**Comment by aidecoe** [comment]
> I'm holding off merging other PRs until this one is one is merged, to prevent further conflicts.

Thank you. That will certainly help!

**Comment by aidecoe** [comment]
hm, I believe I have rebase onto most recent master...

**Comment by aidecoe** [comment]
@marmarek it fails for centos, but in the logs at the end I see "Done. Your build exited with 0.". Could you advise me, please?

**Comment by marmarek** [comment]
A race condition ;)
In the meantime I've restarted it because it failed for some random network issue reason. And now it succeeded.
